[export] Preserve requires_grad for export inputs.

zhxchen17 opened the pull request:
Summary: Today meta['val'] on placeholder nodes doesn't preserve the consistent requires_grad information with the original inputs. Seems there's no easy way to fix this directly at proxy tensor layer. This is useful for reexporting joint graph.

Test Plan: test_preserve_requires_grad_placeholders

Differential Revision: D58555651

tugsbayasgalan commented on test/export/test_export.py:
```diff
@@ -5382,6 +5382,25 @@ def forward(self, w, x, y, z):
                 _disable_forced_specializations=True,
             )
 
+    @testing.expectedFailureSerDer
```
What fails here? Could you file issue?

zhxchen17 replied:
I think requires_grad are not preserved with roundtripping. I can file an issue later.

tugsbayasgalan replied:
Can you leave a message describing what doesn't work then?

tugsbayasgalan commented on torch/export/_trace.py:
```diff
@@ -339,6 +339,41 @@ def _get_param_buffer_mapping(
+            InputKind.BUFFER,
+        ):
+            assert spec.target is not None
+            node.meta["val"].requires_grad = fake_params_buffers[
```
Should we do this at fake tensor constructor? I think fake tensor should reflect faithfully what the original tensor does.

zhxchen17 replied:
we couldn't right now. see comments here: https://fburl.com/code/vgl33i42

tugsbayasgalan replied:
Do we not use the same inputs for "val" field? Interesting...

zhxchen17 commented:
@pytorchbot rebase

zhxchen17 committed:
[export] Preserve requires_grad for export inputs. (#128656)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/128656

Today meta['val'] on placeholder nodes doesn't preserve the consistent requires_grad information with the original inputs. Seems there's no easy way to fix this directly at proxy tensor layer. This is useful for reexporting joint graph.

Test Plan: test_preserve_requires_grad_placeholders

Reviewed By: tugsbayasgalan

Differential Revision: D58555651
